samza-core/src/main/scala/org/apache/samza/system/SystemConsumers.scala

```diff
@@ -208,7 +223,11 @@ class SystemConsumers (
       if (startpoint != null) {
         consumer.register(systemStreamPartition, startpoint)
       } else {
-        consumer.register(systemStreamPartition, offset)
+        val existingOffset = sspToRegisteredOffsets.get(systemStreamPartition)
+        val systemAdmin = systemAdmins.getSystemAdmin(systemStreamPartition.getSystem)
+        if (existingOffset == null || systemAdmin.offsetComparator(existingOffset, offset) > 0) {
+          sspToRegisteredOffsets.put(systemStreamPartition, offset)
```

Comment:
It seems conceptually a bit odd to call `SystemConsumer.register` in `SystemConsumers.start`. If `SystemConsumer.register` works with multiple calls (so it just registers the offset from the latest call to `register`), then you could call `register` here, and then you wouldn't need to call it in `start`.
Not sure if this is feasible, but I think it would make the flow a bit cleaner, since you wouldn't have to worry about doing anything extra in `start`.

Reply:
Adding to @cameronlee314's comment. What's the motivation for only accepting the older offset if `register` is called multiple times? When would this happen? Is it the broadcast stream scenario?

Reply:
1. This would happen in the broadcast stream scenario. This older-offset-comparator check is done by some of the SystemConsumer implementation and not in the others. I'm moving it to a common layer like `SystemConsumers` above to ensure functional correctness. More context [here](https://github.com/apache/samza/pull/918#discussion_r257523873).
2. Currently this odd behavior is what all the system consumer implementations that support broadcast streams do.